### PR TITLE
[move][move-compiler] Add debugging flags and support to the compiler

### PR DESCRIPTION
**Pull request by cgswords**
## Description 

This introduces a `debug_print` macro to the compiler, plus adds some flags in HLIR. These are to enable reusable debugging in a concise, extensible way.

Note that `debug_print` is currently feature-flagged so it only occurs when `debug_assertions` are enabled.

## Test plan 

Manually enabling the flags and running tests locally does the printing, but otherwise it does nothing.

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:

### CODE REVIEW

File: external-crates/move/crates/move-compiler/src/hlir/translate.rs [hunk 1]
```diff
@@ -122,8 +122,18 @@ type VariantFieldIndicies = UniqueMap<
 
 type VariantPositionalMap =
     UniqueMap<ModuleIdent, UniqueMap<DatatypeName, UniqueMap<VariantName, bool>>>;
-pub struct Context<'env> {
+
+pub(super) struct HlirDebugFlags {
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/src/hlir/translate.rs:126]:
```suggestion
pub(super) struct HLIRDebugFlags {
```

Comment (tnowacki) [reply external-crates/move/crates/move-compiler/src/hlir/translate.rs:126]: I think?

File: external-crates/move/crates/move-compiler/src/hlir/translate.rs [hunk 2]
```diff
@@ -277,8 +287,16 @@ impl<'env> Context<'env> {
                 &mdef.enums,
             );
         }
+        let debug = HlirDebugFlags {
+            match_translation: false,
+            match_variant_translation: false,
+            match_specialization: false,
+            match_counterexample: false,
+            match_work_queue: false,
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/src/hlir/translate.rs:295]: should we come up with a way of setting these without recompilation?

**Review by tnowacki — APPROVED** [review]
Neat! I can't wait to start sprinkling things around!

File: external-crates/move/crates/move-compiler/src/hlir/translate.rs (hunk 2 above)
Comment (cgswords) [reply external-crates/move/crates/move-compiler/src/hlir/translate.rs:295]: These could eventually come live as part of the compiler clap flags, but that's user-facing so I thought to err on the side of caution for now.

**Pull request merged**